app/src/main/java/org/mozilla/fenix/searchdialog/SearchDialogFragment.kt

```diff
@@ -324,6 +337,66 @@ class SearchDialogFragment : AppCompatDialogFragment(), UserInteractionHandler {
             })
     }
 
+    override fun onRequestPermissionsResult(
+        requestCode: Int,
+        permissions: Array<String>,
+        grantResults: IntArray
+    ) {
+        when (requestCode) {
+            REQUEST_CODE_CAMERA_PERMISSIONS -> qrFeature.withFeature {
+                context?.let { context: Context ->
+                    it.onPermissionsResult(permissions, grantResults)
+                    if (!context.isPermissionGranted(Manifest.permission.CAMERA)) {
+                        cameraPermissionsDenied = true
+                        dismissAndResetFocus()
+                    }
+                }
+            }
+            else -> super.onRequestPermissionsResult(requestCode, permissions, grantResults)
+        }
+    }
+
+    private fun dismissAndResetFocus() {
+        view?.qr_scan_button?.isChecked = false
+        toolbarView.view.edit.focus()
+        toolbarView.view.requestFocus()
+    }
+
+    private fun showPermissionsNeededDialog() {
+        AlertDialog.Builder(requireContext()).apply {
+            val spannableText = SpannableString(
+                resources.getString(R.string.camera_permissions_needed_message)
+            )
+            setMessage(spannableText)
+            setNegativeButton(R.string.camera_permissions_needed_negative_button_text) {
+                    dialog: DialogInterface, _ ->
+                dialog.cancel()
+                dismissAndResetFocus()
+            }
+            setPositiveButton(R.string.camera_permissions_needed_positive_button_text) {
+                    dialog: DialogInterface, _ ->
+                val intent: Intent = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
+                    val intent = Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS)
+                    val uri = Uri.fromParts("package", requireContext().packageName, null)
```

Comment:
Would benefit from some comments about what's going on here with the URI library